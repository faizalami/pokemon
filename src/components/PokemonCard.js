/** @jsxImportSource @emotion/react */
import styled from '@emotion/styled';
import { margin, padding, rounded, width } from './utilities';
import { Flex } from './FlexGrid';
import { darkGray, gray } from './variables';
import Image from './Image';
import { css } from '@emotion/react';
import { ButtonLink } from './Buttons';

const CardWrapper = styled(Flex)`
  border: 1px solid ${gray};
  ${rounded}
`;

const cardThumbnailStyle = css`
  object-position: center;
  object-fit: contain;
  height: 150px;
  border: 0.5rem solid transparent;
  ${width.full}
`;

const displayInline = css`display: inline-block`;

const PokemonName = styled.h2`
  font-weight: normal;
  text-align: center;
  ${margin.a2}
  ${margin.b1}
`;

const PokemonOriginalName = styled.p`
  text-align: center;
  color: ${darkGray};
  ${margin.a2}
  ${margin.t0}
`;

function PokemonCard ({ id, name, nickname }) {
  return (
    <ButtonLink to={`/pokemon/${name}`} variant="link" css={[padding.a0, displayInline]}>
      <CardWrapper column>
        <Image
          imageCss={cardThumbnailStyle}
          src={`${process.env.REACT_APP_DREAM_WORLD_URL}${id}.svg`}
          alt={name}
          lazy
          width={300}
          height={300}
        />

        <PokemonName>{nickname || name}</PokemonName>
        {nickname ? <PokemonOriginalName>{name}</PokemonOriginalName> : null}
      </CardWrapper>
    </ButtonLink>
  );
}

export default PokemonCard;
