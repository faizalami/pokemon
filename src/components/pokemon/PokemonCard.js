/** @jsxImportSource @emotion/react */
import styled from '@emotion/styled';
import { margin, padding, rounded, width } from '../utilities';
import { Flex } from '../FlexGrid';
import { darkGray, gray, pokeBallDark, pokeBallDarkRed } from '../variables';
import Image from '../Image';
import { css } from '@emotion/react';
import { Button, ButtonLink } from '../Buttons';
import { useState } from 'react';
import ReleaseConfirmModal from './ReleaseConfirmModal';

const CardWrapper = styled(Flex)`
  border: 1px solid ${gray};
  ${rounded}
`;

const cardThumbnailStyle = css`
  position: relative;
  height: 150px;

  .thumbnail-background {
    position: absolute;
    height: 150px;
    border: 0.5rem solid transparent;
    opacity: 0.7;
    ${rounded}
    ${width.full}
  }

  .thumbnail-image {
    position: absolute;
    object-position: center;
    object-fit: contain;
    height: 150px;
    border: 0.5rem solid transparent;
    ${width.full}
  }
`;

const pokemonCountBadge = css`
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
  width: 2rem;
  height: 2rem;

  * {
    position: absolute;
  }

  img {
    width: 2rem;
    height: 2rem;
  }

  .pokemon-total {
    color: ${pokeBallDarkRed};
    background-color: white;
    border-radius: 50%;
    font-size: 0.75rem;
    width: 1.2rem;
    height: 1.2rem;
    margin: 0.4rem;
    border: 1.5px solid ${pokeBallDark};
  }
`;

const cardThumbnailBackground = color => css`background-color: ${color}`;

const displayInline = css`display: inline-block`;

const PokemonName = styled.h2`
  font-weight: normal;
  text-align: center;
  ${margin.a2}
  ${margin.b1}
`;

const PokemonOriginalName = styled.p`
  text-align: center;
  color: ${darkGray};
  ${margin.a2}
  ${margin.t0}
`;

const releaseButtonStyle = css`
  img {
    display: inline-flex;
    width: 1.25rem;
    height: 1.25rem;
    ${margin.t1}
  }

  p {
    color: white;
    display: inline-flex;
    font-size: 1rem;
    ${margin.a0}
    ${margin.l2}
  }
`;

function PokemonTotal ({ total }) {
  if (total) {
    const totalDisplay = total > 9 ? '9+' : total;
    return (
      <div css={pokemonCountBadge}>
        <Image
          png="/assets/img/png/pokeball-64.png"
          webp="/assets/img/webp/pokeball-64.webp"
          alt="Release"
          lazy
          width={24}
          height={24}
        />
        <Flex justifyContent="center" alignItems="center" className="pokemon-total">{totalDisplay}</Flex>
      </div>
    );
  }
  return null;
}

function PokemonCard ({ id, nickname_id, name, nickname, species, pokemonsTotal, lazy }) {
  const [pokemonToDelete, setPokemonToDelete] = useState(null);

  const total = !nickname ? pokemonsTotal?.[id] : null;

  const handleReleaseClick = () => {
    setPokemonToDelete({ nickname_id, nickname });
  };

  return (
    <>
      <CardWrapper column>
        <ButtonLink to={`/pokemon/${name}`} variant="link" css={[padding.a0, displayInline]}>
          <div css={cardThumbnailStyle}>
            <div className="thumbnail-background" css={cardThumbnailBackground(species?.color?.name)}/>
            <Image
              className="thumbnail-image"
              src={`${process.env.REACT_APP_DREAM_WORLD_URL}${id}.svg`}
              alt={name}
              lazy={lazy}
              width={300}
              height={300}
            />
            <PokemonTotal total={total}/>
          </div>

          <PokemonName>{nickname || name}</PokemonName>
          {nickname_id ? <PokemonOriginalName>{name}</PokemonOriginalName> : null}
        </ButtonLink>
        {nickname_id ? (
          <Button css={releaseButtonStyle} onClick={handleReleaseClick}>
            <Image
              png="/assets/img/png/pokeball-release-64.png"
              webp="/assets/img/webp/pokeball-release-64.webp"
              alt="Release"
              lazy
              width={64}
              height={64}
            />
            <p>Release</p>
          </Button>
        ) : null}
      </CardWrapper>
      <ReleaseConfirmModal pokemon={pokemonToDelete} onClose={() => setPokemonToDelete(null)}/>
    </>
  );
}

export default PokemonCard;
