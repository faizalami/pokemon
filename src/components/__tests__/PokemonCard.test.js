import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import PokemonCard from '../PokemonCard';

describe('Test Pokemon Card', () => {
  test('PokemonCard matches snapshot', () => {
    const dummy = {
      id: 1,
      name: 'bulbasaur',
    };

    expect(render(
      <MemoryRouter>
        <PokemonCard {...dummy} />
      </MemoryRouter>,
    )).toMatchSnapshot();
  });

  test('using nickname', () => {
    const dummy = {
      id: 1,
      name: 'bulbasaur',
      nickname: 'test',
    };

    render(
      <MemoryRouter>
        <PokemonCard {...dummy} />
      </MemoryRouter>,
    );

    const displayName = screen.getByText(dummy.nickname);
    const realName = screen.queryByText(dummy.name);
    expect(displayName.tagName).toBe('H2');
    expect(realName).toBeInTheDocument();
  });
});
