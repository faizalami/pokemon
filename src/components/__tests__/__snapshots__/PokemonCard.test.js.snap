// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Pokemon Card PokemonCard matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  border-radius: 0.375rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  cursor: pointer;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background-color: transparent;
  border: 1px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0rem;
  display: inline-block;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  border: 1px solid #d1d5db;
  border-radius: 0.5rem;
}

.emotion-2 {
  position: relative;
  height: 150px;
}

.emotion-2 .thumbnail-background {
  position: absolute;
  background-color: #d1d5db;
  height: 150px;
  border: 0.5rem solid transparent;
  opacity: 0.7;
  border-radius: 0.5rem;
  width: 100%;
}

.emotion-2 img {
  position: absolute;
  object-position: center;
  object-fit: contain;
  height: 150px;
  border: 0.5rem solid transparent;
  width: 100%;
}

.emotion-4 {
  font-weight: normal;
  text-align: center;
  margin: 0.5rem;
  margin-bottom: 0.25rem;
}

<body>
    <div>
      <a
        class="emotion-0"
        href="/pokemon/bulbasaur"
      >
        <div
          class="emotion-1"
        >
          <div
            class="emotion-2"
          >
            <div
              class="thumbnail-background"
            />
            <picture>
              <img
                alt="bulbasaur"
                class="emotion-3"
                data-testid="image-regular"
                height="300"
                loading="lazy"
                src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
                width="300"
              />
            </picture>
          </div>
          <h2
            class="emotion-4"
          >
            bulbasaur
          </h2>
        </div>
      </a>
    </div>
  </body>,
  "container": .emotion-0 {
  border-radius: 0.375rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: 1rem;
  cursor: pointer;
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  padding-left: 1rem;
  padding-right: 1rem;
  background-color: transparent;
  border: 1px solid transparent;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0rem;
  display: inline-block;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  border: 1px solid #d1d5db;
  border-radius: 0.5rem;
}

.emotion-2 {
  position: relative;
  height: 150px;
}

.emotion-2 .thumbnail-background {
  position: absolute;
  background-color: #d1d5db;
  height: 150px;
  border: 0.5rem solid transparent;
  opacity: 0.7;
  border-radius: 0.5rem;
  width: 100%;
}

.emotion-2 img {
  position: absolute;
  object-position: center;
  object-fit: contain;
  height: 150px;
  border: 0.5rem solid transparent;
  width: 100%;
}

.emotion-4 {
  font-weight: normal;
  text-align: center;
  margin: 0.5rem;
  margin-bottom: 0.25rem;
}

<div>
    <a
      class="emotion-0"
      href="/pokemon/bulbasaur"
    >
      <div
        class="emotion-1"
      >
        <div
          class="emotion-2"
        >
          <div
            class="thumbnail-background"
          />
          <picture>
            <img
              alt="bulbasaur"
              class="emotion-3"
              data-testid="image-regular"
              height="300"
              loading="lazy"
              src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/1.svg"
              width="300"
            />
          </picture>
        </div>
        <h2
          class="emotion-4"
        >
          bulbasaur
        </h2>
      </div>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
